Bugsnag/KSCrash/Source/KSCrash/Recording/BSG_Jailbreak.h

```diff
@@ -0,0 +1,215 @@
+    register uintptr_t call asm("x16") = (uintptr_t)(call_num); \
+    register uintptr_t p0 asm("x0") = (uintptr_t)(param0); \
+    register uintptr_t p1 asm("x1") = (uintptr_t)(param1); \
+    register uintptr_t p2 asm("x2") = (uintptr_t)(param2); \
+    register uintptr_t carry_bit asm("x3") = 0; \
+    asm volatile("svc #0x80\n" \
+                 "mov x3, #0\n" \
+                 "adc x3, x3, x3\n" \
+                 : "=r"(carry_bit), "=r"(p0),"=r"(p1) \
+                 : "r"(p0), "r"(p1), "r"(p2), "r"(call), "r"(carry_bit) \
+                 : "memory", "cc"); \
+    if(carry_bit == 1) { \
+        *(pResult) = -1; \
+    } else {\
+        *(pResult) = (int)p0; \
+    } \
+} while(0)
+
+#elif TARGET_CPU_X86_64
+#define BSG_HAS_CUSTOM_SYSCALL 1
+
+// X86_64 3-parameter syscall
+// Writes -1 to *(pResult) on failure, or the actual result on success.
+//
+// Implementation Details:
+// - Syscall# is in RAX, params in RDI, RSI, RDX, and return in RAX.
+// - Carry bit is cleared on success, set on failure.
+// - We must also inform the compiler that memory, rcx, r11 may get clobbered.
+#define bsg_syscall3(call_num, param0, param1, param2, pResult) do { \
+    register uintptr_t rax = (uintptr_t)(call_num) | (2<<24); \
+    register uintptr_t p0 = (uintptr_t)(param0); \
+    register uintptr_t p1 = (uintptr_t)(param1); \
+    register uintptr_t p2 = (uintptr_t)(param2); \
+    register uintptr_t carry_bit = 0; \
+    asm volatile( \
+        "syscall" \
+        : "=@ccc"(carry_bit), "+a"(rax) \
+        : "D" (p0), "S" (p1), "d" (p2) \
+        : "memory", "rcx", "r11"); \
+    if(carry_bit == 1) { \
+        *(pResult) = -1; \
+    } else { \
+        *(pResult) = (int)rax; \
+    } \
+} while(0)
+
+#else
+#define BSG_HAS_CUSTOM_SYSCALL 0
+
+// Unhandled architecture.
+// We fall back to the libc functions in this case, mimicing the syscall-like API.
+// See definitions below.
+
+#endif /* TARGET_CPU_XYZ */
+
+
+#if BSG_HAS_CUSTOM_SYSCALL
+
+// See: https://opensource.apple.com/source/xnu/xnu-7195.81.3/bsd/kern/syscalls.master
+#define BSG_SYSCALL_OPEN 5
+#define bsg_syscall_open(path, flags, mode, pResult) bsg_syscall3(BSG_SYSCALL_OPEN, (uintptr_t)path, flags, mode, pResult)
+
+#else
+
+#define bsg_syscall_open(path, flags, mode, pResult) do {*(pResult) = open(path, flags, mode);} while(0)
+
+#endif /* BSG_HAS_CUSTOM_SYSCALL */
+
+
+/**
+ * Get this device's jailbreak status.
+ * Stores nonzero in *(pIsJailbroken) if the device is jailbroken, 0 otherwise.
+ * Note: Implemented as a macro to force it inline always.
+ */
+#if !TARGET_OS_SIMULATOR && !TARGET_OS_OSX
+#define get_jailbreak_status(pIsJailbroken) do { \
+    int fd = 0; \
+ \
+    bool tmp_file_is_accessible = false; \
+    bool mobile_substrate_exists = false; \
+    bool etc_apt_exists = false; \
+    bool has_insert_libraries = false; \
+ \
+    const char* test_write_file = "/tmp/bugsnag-check.txt"; \
+    remove(test_write_file); \
```

Comment:
Question: is there a reason to prefer `remove()` over `unlink()`?

Reply:
No reason. I generally use `remove` because it's portable, whereas `unlink` isn't.